File: clients/android/NewsBlur/src/com/newsblur/fragment/RenameDialogFragment.java
package com.newsblur.fragment;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Toast;

import com.newsblur.R;
import com.newsblur.databinding.DialogRenameBinding;
import com.newsblur.domain.Feed;
import com.newsblur.util.AppConstants;
import com.newsblur.util.FeedUtils;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class RenameDialogFragment extends DialogFragment {

    @Inject
    FeedUtils feedUtils;

    private static final String FEED = "feed";
    private static final String FOLDER = "folder";
    private static final String FOLDER_NAME = "folder_name";
    private static final String FOLDER_PARENT = "folder_parent";
    private static final String RENAME_TYPE = "rename_type";

    public static RenameDialogFragment newInstance(Feed feed) {
        RenameDialogFragment fragment = new RenameDialogFragment();
        Bundle args = new Bundle();
        args.putSerializable(FEED, feed);
        args.putString(RENAME_TYPE, FEED);
        fragment.setArguments(args);
        return fragment;
    }

    public static RenameDialogFragment newInstance(String folderName, String folderParent) {
        RenameDialogFragment fragment = new RenameDialogFragment();
        Bundle args = new Bundle();
        args.putString(FOLDER_NAME, folderName);
        args.putString(FOLDER_PARENT, folderParent);
        args.putString(RENAME_TYPE, FOLDER);
        fragment.setArguments(args);
        return fragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final Activity activity = getActivity();
        LayoutInflater inflater = LayoutInflater.from(activity);
        View v = inflater.inflate(R.layout.dialog_rename, null);
        final DialogRenameBinding binding = DialogRenameBinding.bind(v);

        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setView(v);
        builder.setNegativeButton(R.string.alert_dialog_cancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                RenameDialogFragment.this.dismiss();
            }
        });
        if (getArguments().getString(RENAME_TYPE).equals(FEED)) {
            final Feed feed = (Feed) getArguments().getSerializable(FEED);
            builder.setTitle(String.format(getResources().getString(R.string.title_rename_feed), feed.title));
            binding.inputName.setText(feed.title);
            builder.setPositiveButton(R.string.feed_name_save, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    feedUtils.renameFeed(activity, feed.feedId, binding.inputName.getText().toString());
                    RenameDialogFragment.this.dismiss();
                }
            });
        } else { // FOLDER
            final String folderName = getArguments().getString(FOLDER_NAME);
            final String folderParentName = getArguments().getString(FOLDER_PARENT);

            builder.setTitle(String.format(getResources().getString(R.string.title_rename_folder), folderName));
            binding.inputName.setText(folderName);

            builder.setPositiveButton(R.string.folder_name_save, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    String newFolderName = binding.inputName.getText().toString();
                    if (TextUtils.isEmpty(newFolderName)) {
                        Toast.makeText(activity, R.string.add_folder_name, Toast.LENGTH_SHORT).show();
                        return;
                    }

                    String inFolder = "";
                    if (!TextUtils.isEmpty(folderParentName) && !folderParentName.equals(AppConstants.ROOT_FOLDER)) {
                        inFolder = folderParentName;
                    }
                    feedUtils.renameFolder(folderName, newFolderName, inFolder, activity);
                    RenameDialogFragment.this.dismiss();
                }
            });
        }

        return builder.create();
    }
}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/fragment/RenameDialogFragment.kt
package com.newsblur.fragment

import android.app.Dialog
import android.os.Bundle
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import androidx.hilt.lifecycle.viewmodel.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.newsblur.R
import com.newsblur.design.LocalNbColors
import com.newsblur.design.NewsBlurTheme
import com.newsblur.design.toVariant
import com.newsblur.preference.PrefsRepo
import com.newsblur.viewModel.RenameDialogViewModel
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class RenameDialogFragment : DialogFragment() {
    @Inject
    lateinit var prefsRepo: PrefsRepo

    companion object {
        private const val RENAME_TYPE = "rename_type"
        private const val TYPE_FEED = "feed"
        private const val TYPE_FOLDER = "folder"

        private const val FEED_ID = "feed_id"
        private const val FEED_TITLE = "feed_title"

        private const val FOLDER_NAME = "folder_name"
        private const val FOLDER_PARENT = "folder_parent"

        @JvmStatic
        fun newFeedInstance(
            feedId: String,
            feedTitle: String,
        ) = RenameDialogFragment().apply {
            arguments =
                bundleOf(
                    RENAME_TYPE to TYPE_FEED,
                    FEED_ID to feedId,
                    FEED_TITLE to feedTitle,
                )
        }

        @JvmStatic
        fun newFolderInstance(
            folderName: String,
            folderParent: String?,
        ) = RenameDialogFragment().apply {
            arguments =
                bundleOf(
                    RENAME_TYPE to TYPE_FOLDER,
                    FOLDER_NAME to folderName,
                    FOLDER_PARENT to folderParent,
                )
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val type = requireArguments().getString(RENAME_TYPE)!!

        val composeView =
            ComposeView(requireContext()).apply {
                setContent {
                    NewsBlurTheme(variant = prefsRepo.getSelectedTheme().toVariant()) {
                        when (type) {
                            TYPE_FEED -> {
                                val feedId = requireArguments().getString(FEED_ID)!!
                                val feedTitle = requireArguments().getString(FEED_TITLE)!!
                                RenameDialogContent(
                                    type = RenameType.Feed(feedId = feedId, currentTitle = feedTitle),
                                    onDismiss = { dismissAllowingStateLoss() },
                                    onDone = { dismissAllowingStateLoss() },
                                )
                            }

                            else -> {
                                val folderName = requireArguments().getString(FOLDER_NAME)!!
                                val folderParent = requireArguments().getString(FOLDER_PARENT)
                                RenameDialogContent(
                                    type =
                                        RenameType.Folder(
                                            currentName = folderName,
                                            parent = folderParent,
                                        ),
                                    onDismiss = { dismissAllowingStateLoss() },
                                    onDone = { dismissAllowingStateLoss() },
                                )
                            }
                        }
                    }
                }
            }

        return MaterialAlertDialogBuilder(requireContext())
            .setView(composeView)
            .create()
    }
}

sealed interface RenameType {
    data class Feed(
        val feedId: String,
        val currentTitle: String,
    ) : RenameType

    data class Folder(
        val currentName: String,
        val parent: String?,
    ) : RenameType
}

@Composable
fun RenameDialogContent(
    viewModel: RenameDialogViewModel = hiltViewModel(),
    type: RenameType,
    onDismiss: () -> Unit,
    onDone: () -> Unit,
) {
    val state by viewModel.uiState.collectAsStateWithLifecycle()
    val colors = LocalNbColors.current

    val entityTitle =
        when (type) {
            is RenameType.Feed -> type.currentTitle
            is RenameType.Folder -> type.currentName
        }

    LaunchedEffect(type) {
        when (type) {
            is RenameType.Feed -> viewModel.setInitialText(type.currentTitle)
            is RenameType.Folder -> viewModel.setInitialText(type.currentName)
        }
    }

    LaunchedEffect(state) {
        if (state is RenameDialogViewModel.UiState.Done) onDone()
    }

    AlertDialog(
        onDismissRequest = onDismiss,
        shape = RoundedCornerShape(12.dp),
        tonalElevation = 6.dp,
        containerColor = colors.itemBackground,
        iconContentColor = MaterialTheme.colorScheme.onSurfaceVariant,
        titleContentColor = colors.textDefault,
        textContentColor = colors.textDefault,
        title = {
            val titleRes =
                when (type) {
                    is RenameType.Feed -> R.string.title_rename_feed
                    is RenameType.Folder -> R.string.title_rename_folder
                }
            Text(
                text = stringResource(titleRes, entityTitle),
                style = MaterialTheme.typography.titleMedium,
            )
        },
        text = {
            Column(
                modifier = Modifier.fillMaxWidth(),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                if (state is RenameDialogViewModel.UiState.Loading) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.Center,
                    ) {
                        Text(
                            text =
                                when (type) {
                                    is RenameType.Feed -> stringResource(R.string.renaming_feed_message, entityTitle)
                                    is RenameType.Folder -> stringResource(R.string.renaming_folder_message, entityTitle)
                                },
                            style = MaterialTheme.typography.bodyLarge,
                        )
                        Spacer(Modifier.width(12.dp))
                        CircularProgressIndicator(modifier = Modifier.size(24.dp))
                    }
                } else {
                    RenameTextField(
                        value = viewModel.text,
                        onValueChange = viewModel::onTextChanged,
                        isError = state is RenameDialogViewModel.UiState.ValidationError,
                    )

                    if (state is RenameDialogViewModel.UiState.ValidationError) {
                        Spacer(Modifier.height(8.dp))
                        Text(
                            text = stringResource(R.string.add_folder_name),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.error,
                        )
                    }
                }
            }
        },
        confirmButton = {
            if (state !is RenameDialogViewModel.UiState.Loading) {
                TextButton(
                    onClick = {
                        when (type) {
                            is RenameType.Feed ->
                                viewModel.renameFeed(type.feedId)

                            is RenameType.Folder ->
                                viewModel.renameFolder(type.currentName, type.parent)
                        }
                    },
                ) {
                    Text(
                        text = stringResource(R.string.rename),
                        color = colors.textDefault,
                    )
                }
            }
        },
        dismissButton = {
            if (state !is RenameDialogViewModel.UiState.Loading) {
                TextButton(onClick = onDismiss) {
                    Text(
                        text = stringResource(id = R.string.alert_dialog_cancel),
                        color = colors.textDefault,
                    )
                }
            }
        },
    )
}

@Composable
private fun RenameTextField(
    value: String,
    onValueChange: (String) -> Unit,
    hintRes: Int = R.string.new_folder_name_hint,
    modifier: Modifier = Modifier,
    isError: Boolean = false,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .padding(horizontal = 10.dp),
    ) {
        Spacer(Modifier.height(5.dp))

        OutlinedTextField(
            value = value,
            onValueChange = onValueChange,
            modifier = Modifier.fillMaxWidth(),
            singleLine = true,
            textStyle = MaterialTheme.typography.bodyLarge,
            placeholder = { Text(stringResource(hintRes)) },
            isError = isError,
            keyboardOptions =
                KeyboardOptions(
                    capitalization = KeyboardCapitalization.Sentences,
                    imeAction = ImeAction.Done,
                ),
        )

        Spacer(Modifier.height(10.dp))
    }
}
